src/router_core/modules/mobile_sync/mobile.c

```diff
@@ -0,0 +1,860 @@
+    }
+}
+
+
+static bool qcm_mobile_sync_addr_is_mobile(qdr_address_t *addr)
+{
+    const char *hash_key = (const char*) qd_hash_key_by_handle(addr->hash_handle);
+    return !!strchr("MCDEFH", hash_key[0]);
+}
+
+
+qdr_node_t *qdc_mobile_sync_router_by_id(qdrm_mobile_sync_t *msync, qd_parsed_field_t *id_field)
+{
+    qd_iterator_t *id_iter = qd_parse_raw(id_field);
+    qdr_node_t *router = DEQ_HEAD(msync->core->routers);
+    while (!!router) {
+        if (qd_iterator_equal(id_iter, qd_hash_key_by_handle(router->owning_addr->hash_handle) + 1))
+            return router;
+        router = DEQ_NEXT(router);
+    }
+
+    return 0;
+}
+
+
+/**
+ * Set the 'block_deletion' flag on the address to ensure it is not deleted out from under
+ * our list.  If the flag was already set, make note of that fact so we don't clear it later.
+ */
+static void qcm_mobile_sync_address_added_to_list(qdr_address_t *addr)
+{
+    if (addr->block_deletion) {
+        BIT_SET(addr->sync_mask, ADDR_SYNC_DELETION_WAS_BLOCKED);
+    } else {
+        addr->block_deletion = true;
+    }
+}
+
+
+/**
+ * Clear the 'block_deletion' flag on the address if it was set by this module.
+ * Check the address to have it deleted if it is no longer referenced anywhere.
+ */
+static void qcm_mobile_sync_address_removed_from_list(qdr_core_t *core, qdr_address_t *addr)
+{
+    if (!BIT_IS_SET(addr->sync_mask, ADDR_SYNC_DELETION_WAS_BLOCKED)) {
+        addr->block_deletion = false;
+        qdr_check_addr_CT(core, addr);
+    } else {
+        BIT_CLEAR(addr->sync_mask, ADDR_SYNC_DELETION_WAS_BLOCKED);
+    }
+}
+
+
+static qd_composed_field_t *qcm_mobile_sync_message_headers(const char *address, const char *opcode)
+{
+    qd_composed_field_t *field = qd_compose(QD_PERFORMATIVE_HEADER, 0);
+    qd_compose_start_list(field);
+    qd_compose_insert_bool(field, 0); // durable
+    qd_compose_end_list(field);
+
+    field = qd_compose(QD_PERFORMATIVE_PROPERTIES, field);
+    qd_compose_start_list(field);
+    qd_compose_insert_null(field);            // message-id
+    qd_compose_insert_null(field);            // user-id
+    qd_compose_insert_string(field, address); // to
+    qd_compose_end_list(field);
+
+    field = qd_compose(QD_PERFORMATIVE_APPLICATION_PROPERTIES, field);
+    qd_compose_start_map(field);
+    qd_compose_insert_symbol(field, OPCODE);
+    qd_compose_insert_string(field, opcode);
+    qd_compose_end_map(field);
+
+    return field;
+}
+
+
+static void qcm_mobile_sync_compose_diff_addr_list(qdrm_mobile_sync_t *msync, qd_composed_field_t *field, bool is_added)
+{
+    qdr_address_list_t *list = is_added ? &msync->added_addrs : &msync->deleted_addrs;
+
+    qd_compose_start_list(field);
+    qdr_address_t *addr = DEQ_HEAD(*list);
+    while (addr) {
+        const char *hash_key = (const char*) qd_hash_key_by_handle(addr->hash_handle);
+        qd_compose_insert_string(field, hash_key);
+        if (is_added) {
```

Comment:
Just a suggestion:  you _could_ move this check out of the address loop by setting the list name (SYNC_ADD/DEL) and bit in locals before the loop (like you set the list var).